### ISSUE TITLE
How to add focal loss in darknet？

### ISSUE THREAD

**Issue by TaihuLight** [AlexeyAB/darknet#424]
I find the referenc codes in http://blog.csdn.net/linmingan/article/details/77885832， but I can not successfully implement it in darknet.

**Comment by AlexeyAB** [comment]
* You could not change the source code?
* Or did you implement focal-loss, but it gives you low mAP (precision)?

I added code from your link, **but I didn't test it**: http://blog.csdn.net/linmingan/article/details/77885832
**To use this implementation of focal-loss**, just add `focal_loss=1` in the `[region]` layer in your cfg-file: https://github.com/AlexeyAB/darknet/blob/5e448c00f32e80e13eead72318047ce8368aac9a/cfg/yolo-voc.2.0.cfg#L236

So you can try to fine-tune ResNet-152 with `focal_loss=1` which I added:  https://github.com/AlexeyAB/darknet/blob/master/build/darknet/x64/resnet152_yolo.cfg
* Just download: https://pjreddie.com/media/files/resnet152.weights
* do `darknet.exe partial cfg/resnet152.cfg resnet152.weights resnet152.200 200`
* train `darknet.exe detector train voc.data resnet152_yolo.cfg resnet152.200`

The model will be fine-tuned starting from this layer: https://github.com/AlexeyAB/darknet/blob/0c07837397e791c750e81857e214d315ea24199f/build/darknet/x64/resnet152_yolo.cfg#L1431-L1434

**Comment by AlexeyAB** [comment]
Implementation uses: alpha=0.5, gamma=2

Added code: https://github.com/AlexeyAB/darknet/blob/5e448c00f32e80e13eead72318047ce8368aac9a/src/region_layer.c#L133-L146

In general it should increase mAP, but may be just using focal-loss is not enough to increase mAP. You can see all about RetinaNet: https://arxiv.org/abs/1708.02002v2

As said for RetinaNet in Box - Dev: http://cocodataset.org/#detections-leaderboard

> RetinaNet is a dense one-stage object detector trained with **focal loss** to address the extreme class imbalance in training examples [1]. The **model was fine-tuned with a ResNet-101 pretrained model** on ImageNet-1k. This entry reports single model accuracy with vanilla training and testing procedures without any bells and whistles and runs at 5FPS. [1] Focal Loss for Dense Object Detection

**Comment by VanitarNordic** [comment]
This focal loss only can be applied to the ResNet? can we test it with Darknet-19?

**Comment by AlexeyAB** [comment]
@VanitarNordic It can be applied to **any Yolo networks**. 
In the article they got optimal (precision-speed) result on ResNet-101.

**Comment by VanitarNordic** [comment]
@AlexeyAB 

This focal loss is something like making hard examples to make a better weights?

By the way we can just pas your added CFG parameter and train and see if it will increase the performance. I briefly read the paper, mAP increase was fantastic!

**Comment by AlexeyAB** [comment]
@VanitarNordic Yes, just add `focal_loss=1` to the `[region]` layer in any yolo cfg-file. And update code from this repo.

Also if it leads to Nan, then decrease this value to `0.25` https://github.com/AlexeyAB/darknet/blob/5e448c00f32e80e13eead72318047ce8368aac9a/src/region_layer.c#L135

But I didn't test focal_loss yet. 
And I did not check whether this is the correct implementation: http://blog.csdn.net/linmingan/article/details/77885832

**Comment by AlexeyAB** [comment]
> This focal loss is something like making hard examples to make a better weights?

Yes!

![1_hdqjfvxx1atqmfkifnnyzg](https://user-images.githubusercontent.com/4096485/37038603-493c44c4-2166-11e8-85dd-a3cbfe2f11de.png)

**Comment by VanitarNordic** [comment]
I trained with `focal_loss=1`

I did not face any `nan` problem and I trained till the end for one class and 2,250 iterations. 

I can not say it has improved or decreased the performance because before applying this update I was getting `FP = 0 and FN = 1`, but after this update I get `FP=1 and FN=0`. mAP reduced about 1%. 

it worth to investigate more on a bigger dataset with higher number of classes.

**Comment by TaihuLight** [comment]
@AlexeyAB 
According to the review about the correct implementation: http://blog.csdn.net/linmingan/article/details/77885832  it may be wrong, we need to read the paper or see the correct implementation like https://github.com/marvis/pytorch-yolo2 https://github.com/clcarwin/focal_loss_pytorch  they are tested on VOC2007 and get the better mAP.

**Comment by litingsjj** [comment]
@AlexeyAB Have you tried OHEM (Online Hard Example Mining) in yolo? When I choose negative sample randomly. The background is studied not well.

**Comment by AlexeyAB** [comment]
@litingsjj No, I haven't.

**Comment by spaul13** [comment]
How to apply focal loss for a classification model (in the cfg file), if I don't have the yolo layer?

If I can't then can I include yolo layer before the last layer for any classification model to add the focal loss?
